### PR TITLE
test: added test to validate the LVMO resources

### PR DESCRIPTION
**Pull request by riya-singhal31**
This test validates all the resources created by LVMO

Signed-off-by: riya-singhal31 <riyasinghalji@gmail.com>

### CODE REVIEW

File: e2e/validation.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	Eventually(func() bool {
+		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: csiDriverName, Namespace: csiDriverNamespace}, &cd)
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
+	return nil
+}
+
+// function to validate Daemonset(VG Manager resource).
+func ValidateDaemonset() error {
```
Comment (nbalacha) [inline e2e/validation.go:62]:
```suggestion
func ValidateTopolvmNode() error {
```

Comment (nbalacha) [reply e2e/validation.go:62]: The vgmanager daemonset also needs to be checked.

File: e2e/validation.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	Eventually(func() bool {
+		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: daemonSetName, Namespace: daemonSetNamespace}, &ds)
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
+	return nil
+}
+
+// function to validate TopoLVM Deployment.
+func ValidateDeployment() error {
```
Comment (nbalacha) [inline e2e/validation.go:76]:
```suggestion
func ValidateTopolvmController() error {
```

File: e2e/validation.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	appsv1 "k8s.io/api/apps/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	timeout  = time.Second * 10
+	interval = time.Millisecond * 100
+)
+
```
Comment (nbalacha) [inline e2e/validation.go:18]: Declare all resource names as constants here.

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: e2e/lvm/lvm_suite_test.go [hunk 4]
```diff
@@ -25,10 +24,12 @@ var _ = AfterSuite(func() {
 	tests.AfterTestSuiteCleanup()
 })
 
-var _ = Describe("lvmtest", func() {
-	Context("Run a dummy test", func() {
+// Test to validate all the resources created by LVMO.
+var _ = Describe("Validation test", func() {
+	Context("Validate all the resources", func() {
 		It("Should do nothing", func() {
```
[the remark this answers is not in the document]
Comment (riya-singhal31) [reply e2e/lvm/lvm_suite_test.go]: Updated

File: e2e/validation.go (hunk 1 above)
Comment (riya-singhal31) [reply e2e/validation.go]: Updated

File: e2e/validation.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	timeout                = time.Second * 10
+	interval               = time.Millisecond * 100
+	lvmVolumeGroupName     = "vg1"
+	storageClassName       = "odf-lvm-vg1"
+	csiDriverName          = "topolvm.cybozu.com"
+	daemonSetName          = "topolvm-node"
```
Comment (nbalacha) [inline e2e/validation.go:20]:
```suggestion
	topolvmNodeDaemonSetName          = "topolvm-node"
```

File: e2e/validation.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	deploymentName         = "topolvm-controller"
```
Comment (nbalacha) [inline e2e/validation.go:21]:
```suggestion
	topolvmCtrlDeploymentName         = "topolvm-controller"
```

File: e2e/validation.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+	. "github.com/onsi/gomega"
+	lvmv1alpha1 "github.com/red-hat-storage/lvm-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	timeout                   = time.Second * 10
```
[the remark this answers is not in the document]
Comment (riya-singhal31) [reply e2e/validation.go:15]: It's working fine for 10 s. Should I still change it to 30?

File: e2e/validation.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+// function to validate TopoLVM node.
+func ValidateTopolvmNode() error {
+	ds := appsv1.DaemonSet{}
+
+	Eventually(func() bool {
+		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: topolvmNodeDaemonSetName, Namespace: InstallNamespace}, &ds)
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
```
Comment (nbalacha) [inline e2e/validation.go:69]: Please check that the pod is in ready state as well.

File: e2e/validation.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+	debug("depoloyment found\n")
+	return nil
+}
+
+// Validate all the resources created by LVMO.
+func ValidateResources() error {
+
+	// Validate Storage class
+	err := ValidateStorageClass()
```
Comment (nbalacha) [inline e2e/validation.go:137]: Change the order here to match the order in the LVMCluster reconcile  function:

1. CSI Driver
2. Topolvm controller
3. vg manager
4. LVMVolumeGroup
5. topolvm node
6. Storageclass


```

	// Validate CSI Driver
	err := ValidateCSIDriver()
	if err != nil {
		return err
	}

	//Validate TopoLVM Controller
	err = ValidateTopolvmController()
	if err != nil {
		return err
	}

	// Validate VG Manager Daemonset
	err = ValidateVGManager()
	if err != nil {
		return err
	}
	// Validate LVMVg
	err = ValidateLVMvg()
	if err != nil {
		return err
	}

	// Validate Topolvm node
	err = ValidateTopolvmNode()
	if err != nil {
		return err
	}
	// Validate Storage class
	err = ValidateStorageClass()
	if err != nil {
		return err
	}
```

File: e2e/validation.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	ds := appsv1.DaemonSet{}
+	Eventually(func() bool {
+		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: topolvmNodeDaemonSetName, Namespace: InstallNamespace}, &ds)
+		if err != nil {
+			debug("TopoLVM Node: %v", err)
+			return false
+		}
+		return true
+	}, timeout, interval).Should(BeTrue())
+	debug("TopoLVM node found\n")
```
Comment (nbalacha) [inline e2e/validation.go:87]:
```suggestion
func ValidateTopolvmNode() error {
	ds := appsv1.DaemonSet{}
	Eventually(func() bool {
		debug("%s\n", "Starting function - topolvmnode")
		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: topolvmNodeDaemonSetName, Namespace: InstallNamespace}, &ds)
		if err != nil {
			debug("topolvmNode : %s\n", err.Error())
			return false
		}
		return ds.Status.DesiredNumberScheduled == ds.Status.NumberReady
	}, timeout, interval).Should(BeTrue())
```

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: e2e/validation.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+// function to validate TopoLVM Deployment.
+func ValidateTopolvmController() error {
+	dep := appsv1.Deployment{}
+
+	Eventually(func() bool {
+		err := DeployManagerObj.GetCrClient().Get(context.TODO(), types.NamespacedName{Name: topolvmCtrlDeploymentName, Namespace: InstallNamespace}, &dep)
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
+	debug("depoloyment found\n")
```
Comment (nbalacha) [inline e2e/validation.go:109]:
```suggestion
	debug("topolvm-controller deployment found\n")
```

Comment (riya-singhal31) [reply e2e/validation.go]: Updated

**Review by nbalacha — APPROVED** [review]

**Pull request merged**